### ISSUE TITLE
NativeAOT to improve performance for formatting a single file

### ISSUE THREAD

**Issue by belav** [belav/csharpier#474]
NativeAOT may be a solution to the problem of csharpier being slow to format a single file.
It will be in .net7
https://github.com/dotnet/runtime/issues/61231
It is currently experimental, and I believe can be tested now to see if it would solve the problem
https://github.com/dotnet/runtimelab/tree/feature/NativeAOT#readme

**Comment by belav** [comment]
Followed the instructions at https://github.com/dotnet/runtimelab/blob/feature/NativeAOT/docs/using-nativeaot/compiling.md

![image](https://user-images.githubusercontent.com/3606529/141329965-0f0abf9e-6fd4-48c2-af92-818480be013a.png)

**Comment by OneCyrus** [comment]
now that .NET 7 is out, are there some technical blockers from moving to native aot? just wondering about the current state

**Comment by belav** [comment]
I don't believe it is possible to publish a dotnet tool with NativeAOT.
[ReadyToRun](https://learn.microsoft.com/en-us/dotnet/core/deploying/ready-to-run) is another possibility to improve startup time that I just saw. There is an open issue around support for dotnet tools and ReadyToRun https://github.com/dotnet/sdk/issues/10801, but I haven't seen one for NativeAOT.

**Comment by belav** [comment]
ChatGPT seems to think that it would be possible to publish to nuget with platform specific runtimes that use nativeAOT, I should dig into that more

```
Publishing a .NET tool that uses **NativeAOT** involves additional considerations, as tools are typically installed globally and need to support multiple platforms seamlessly. Here’s how you can approach it:

---

### Steps to Publish a .NET Tool with NativeAOT

#### 1. **Set Up NativeAOT Compilation**
   - Ensure your project is configured for NativeAOT in the `.csproj` file:
     ```xml
     <PropertyGroup>
       <PublishAot>true</PublishAot>
       <SelfContained>true</SelfContained>
       <RuntimeIdentifier>win-x64</RuntimeIdentifier> <!-- Or other RIDs -->
     </PropertyGroup>
     ```
   - Replace `RuntimeIdentifier` with the platform(s) you intend to target, e.g., `linux-x64`, `osx-arm64`, etc.

---

#### 2. **Multi-Runtime Support**
   - Tools compiled with NativeAOT are platform-specific. If your tool needs to support multiple platforms, you must build separate binaries for each platform and package them together.
   - Use the `runtimes` folder structure in your package:
     ```
     /runtimes
       /win-x64
         /native
           tool.exe
       /linux-x64
         /native
           tool
       /osx-arm64
         /native
           tool
     ```

---


#### 4. **Create a Multi-Platform NuGet Package**
   - Include binaries for all target platforms in the NuGet package.
   - Configure the `.nuspec` file or `.csproj` to include platform-specific binaries:
     ```xml
     <ItemGroup>
       <None Include="runtimes\win-x64\native\*" Pack="true" PackagePath="runtimes/win-x64/native/" />
       <None Include="runtimes\linux-x64\native\*" Pack="true" PackagePath="runtimes/linux-x64/native/" />
       <None Include="runtimes\osx-arm64\native\*" Pack="true" PackagePath="runtimes/osx-arm64/native/" />
     </ItemGroup>
     ```

---

```

**Comment by belav** [comment]
ChatGpt was hallucinating, here is an issue for adding AOT support in dotnet tools - https://github.com/dotnet/sdk/issues/40931

**Comment by belav** [comment]
This will now be possible with net10

https://github.com/baronfel/multi-rid-tool

**Comment by belav** [comment]
A good article about publishing net10 tools - https://andrewlock.net/exploring-dotnet-10-preview-features-7-packaging-self-contained-and-native-aot-dotnet-tools-for-nuget/

**Comment by SapiensAnatis** [comment]
Looked into this briefly, the main problem with this is that NativeAOT tools only work with the .NET 10 SDK. 

You can provide a runtime-dependent fallback (i.e. a standard .NET tool that runs in CoreCLR) as per https://learn.microsoft.com/en-us/dotnet/core/tools/rid-specific-tools#provide-a-framework-dependent-version, but this feature seems to mainly be about not providing native binaries for niche platforms like `win-arm64` or `linux-musl-x64`. If you try to install the `dotnet10-hybrid-tool` they made for the example from the web using SDK 8.0 or 9.0 it will still fail.

In other words, we could do this but it would require CSharpier to take a hard dependency on the .NET 10 SDK.

Also worth considering is if we have to build for multiple platforms separately then that will need some multiple of the current CI time, likely including macOS arm64 runners which I believe have more aggressive usage caps / cost a lot more than Linux or Windows runners.
